### ISSUE TITLE
IntegrityError: UNIQUE constraint failed: lnschema_bionty_organism.ontology_id

### ISSUE THREAD

**Issue by falexwolf** [laminlabs/lamindb#1353]
Can somehow be produced during transfer.

A file uid from the cellxgene instance to inspect: `1ks6CDWlF2r3GFWGHzz3`.

**Comment by jkobject** [comment]
yet another error on another file: when doing file.save() (on a cellxgene file)

![Screenshot 2023-12-13 at 13 23 32](https://github.com/laminlabs/lamindb/assets/17239533/47801ac1-8aa1-45aa-bce6-27a77d7245ab)
![Screenshot 2023-12-13 at 13 23 56](https://github.com/laminlabs/lamindb/assets/17239533/c220bee9-8d44-40e7-befe-e8adc7b5492e)

**Comment by jkobject** [comment]
actually it is even when doing just `cx_dataset.files.all()`

**Comment by falexwolf** [comment]
Thank you, @jkobject - @sunnyosun & I talked through these things this morning, and @sunnyosun is working on the latest more comprehensive release. She should be able to tell if some files are somehow corrupted.

> cx_dataset.files.all()

These here runs on CI on the whole dataset and should never fail - maybe it's because of the lamindb 0.64 migration of that instance?

**Comment by falexwolf** [comment]
@sunnyosun's and my main current theory is that you're trying to save the same new entity records from multiple threads.

Is this possible?
<img width="489" alt="image" src="https://github.com/laminlabs/lamindb/assets/16916678/ef8883da-f1e2-4e31-a42b-4dfbeedbd70d">

**Comment by falexwolf** [comment]
We do have a problem with records in the source database that are not yet in the target database and are written by multiple threads at the same time. The current check is based on looking at what's in the target database, and hence, doesn't anticipate by a clash coming from two threads that try to write the same thing.

We'll think about this. There should be a solution to make this check thread-safe.

**Comment by jkobject** [comment]
I am not running it in parallel if that is the question

**Comment by jkobject** [comment]
Now I also just got a new error for again saving a file from cellxgene (uid=tQaZHQ9Lyy5AE2n5RC2x):

`IndexError: list index out of range`

![Screenshot 2023-12-13 at 15 34 37](https://github.com/laminlabs/lamindb/assets/17239533/a8c166eb-dfc5-425a-b6f1-795bf7bcfab1)

![Screenshot 2023-12-13 at 15 34 43](https://github.com/laminlabs/lamindb/assets/17239533/ef582976-9ca3-4d38-8889-877009409a6e)

**Comment by sunnyosun** [comment]
Could you print out your Organism registry here? `lb.Organism.df()`

**Comment by jkobject** [comment]
![Screenshot 2023-12-13 at 17 29 34](https://github.com/laminlabs/lamindb/assets/17239533/4ecedc09-b054-45cc-a109-8987a28c787e)

**Comment by sunnyosun** [comment]
Could you try again? (There was an inconsistency of uid for "mouse" due to the update of the cellxgene, I fixed it now.)

**Comment by jkobject** [comment]
I will. I also got this error now for "uid=zwom62bNy5XaeQRF7wc1"

when trying to load the file:

```
File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/h5py/_hl/group.py:499, in Group.__iter__(self)
    496 @with_phil
    497 def __iter__(self):
    498     """ Iterate over member names """
--> 499     for x in self.id.__iter__():
    500         yield self._d(x)

File h5py/h5g.pyx:469, in h5py.h5g.GroupID.__iter__()

File h5py/h5g.pyx:470, in h5py.h5g.GroupID.__iter__()

File h5py/h5g.pyx:102, in h5py.h5g.GroupIter.__init__()

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File h5py/h5g.pyx:336, in h5py.h5g.GroupID.get_num_objs()

RuntimeError: Unable to get group info (bad symbol table node signature)
```

**Comment by sunnyosun** [comment]
Just tried this file and it worked without any issue for me.
<img width="846" alt="Screenshot 2023-12-13 at 22 30 38" src="https://github.com/laminlabs/lamindb/assets/38218185/fc44b5ac-2614-4352-9ad0-5ce9a7a36fd9">

Perhaps you need to update your `anndata` or `p5py` version? I'm using `0.10.2` for `anndata`, and `3.8.0` for `h5py`.

@Koncopd Do you have any idea what it could be?

**Comment by jkobject** [comment]
Another issue popped for uid=vWr4zjODRjjueCITy18e:

This time at the file.save() step:
```
-------------------------------------------------------------------
IndexError                        Traceback (most recent call last)
Cell In[31], line 1
----> 1 preprocessed_dataset = do_preprocess(cx_dataset, start_at=3
3)                                                                 

File ~/Documents code/scPRINT/scprint/dataset/preprocess.py:156, in
 Preprocessor.__call__(self, data, name, description, start_at)
    154         raise v
    155 try:
--> 156     file.save()
    157 except IntegrityError as e:
    158     # UNIQUE constraint failed: lnschema_bionty_organism.on
tology_id
    159     print(f"seeing {e}... continuing")

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb
/_artifact.py:975, in save(self, *args, **kwargs)
    974 def save(self, *args, **kwargs) -> None:
--> 975     self._save_skip_storage(*args, **kwargs)
    976     from lamindb._save import check_and_attempt_clearing, c
heck_and_attempt_upload
    978     exception = check_and_attempt_upload(self)

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb
/_artifact.py:989, in _save_skip_storage(file, *args, **kwargs)
    987 def _save_skip_storage(file, *args, **kwargs) -> None:
    988     save_feature_sets(file)
--> 989     super(Artifact, file).save(*args, **kwargs)
    990     save_feature_set_links(file)

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb
/_registry.py:476, in save(self, *args, **kwargs)
    474 self.features._add_from(self_on_db)
    475 logger.info("transfer labels")
--> 476 self.labels.add_from(self_on_db)

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb/dev/_label_manager.py:177, in LabelManager.add_from(self, data)
    173                 # labels records from data db
    174                 labels = data.labels.get(
    175                     getattr(features_lookup_data, row["name"]), mute=True
    176                 )
--> 177                 transfer_add_labels(labels, features_lookup_self, self, row)                                                  
    179 # for now, have this be duplicated, need to disentangle above
    180 for related_name, (_, labels) in get_labels_as_dict(data).items():

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb/dev/_label_manager.py:67, in transfer_add_labels(labels, features_lookup_self, self, row)
     61     self._host.labels.add(
     62         validated_labels + new_labels,
     63         feature=getattr(features_lookup_self, row["name"]),
     64     )
     66 # validate labels on the default db
---> 67 result = validate_labels(labels)
     68 if isinstance(result, Dict):
     69     for _, (validated_labels, new_labels) in result.items():

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb/dev/_label_manager.py:93, in validate_labels(labels)
     91         result[registry] = validate_labels_registry(labels_registry)  
     92 else:
---> 93     return validate_labels_registry(labels)

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/lamindb/dev/_label_manager.py:77, in validate_labels.<locals>.validate_labels_registry(labels)
     76 def validate_labels_registry(labels: Union[QuerySet, List, Dict]):
---> 77     registry = labels[0].__class__
     78     field = REGISTRY_UNIQUE_FIELD.get(registry.__name__.lower(), "uid")
     79     label_uids = np.array(
     80         [getattr(label, field) for label in labels if label is not None]
     81     )

File ~/miniconda3/envs/scprint/lib/python3.10/site-packages/django/db/models/query.py:450, in QuerySet.__getitem__(self, k)
    448 qs.query.set_limits(k, k + 1)
    449 qs._fetch_all()
--> 450 return qs._result_cache[0]

IndexError: list index out of range
```

**Comment by jkobject** [comment]
in your usecase, does it also work when you are doing the same thing as I do:
`backed().obs. is_primary_data.sum()`

**Comment by sunnyosun** [comment]
Also no issues:
<img width="536" alt="Screenshot 2023-12-14 at 09 03 23" src="https://github.com/laminlabs/lamindb/assets/38218185/37475e5e-e02e-416f-a31c-bd9640c46aeb">

**Comment by sunnyosun** [comment]
If you are using `.load`, it is the same as running `anndata.read(artifact.stage())`, which is likely unrelated to LaminDB. I wonder if your downloaded file is somehow corrupted. Maybe you can try deleting the cached file at path `artifact.stage()` and try again.

**Comment by sunnyosun** [comment]
The error says it somehow think you have a local version of this file but couldn't find it. Could you try to download it with `.stage()`?

**Comment by jkobject** [comment]
Could it be an issue with files I have downloaded / saved a few weeks ago but then incorrectly removed. 

even when getting them from the cellxgene instance, lamin sees that they are saved in my local instance but then they are corrupted in some way. For example here, the path to the file actually doesn't exist anymore.

**Comment by jkobject** [comment]
the .stage() just gives the path that doesn't exists

**Comment by sunnyosun** [comment]
And `.stage()` doesn't start downloading the file?

**Comment by jkobject** [comment]
Nope. but for all files where I got an issue, I will see if removing them from my instance completely and then redownloading them from the cellxgene instance solves the issue

**Comment by jkobject** [comment]
ok so I still got the issue for `r4iCehg3Tw5IbCLirSog`.

However getting the file from `ln.Artifact.filter() `and then saving it, works. but the same thing from cellxgene's instance doesn't `ln.Dataset.using("laminlabs/cellxgene").filter().first()`

**Comment by sunnyosun** [comment]
Can you share the full traceback of your commands? Are you now having issues with even filtering? `ln.Dataset.using("laminlabs/cellxgene").filter().first()` works for me.

**Comment by Koncopd** [comment]
Ok, lets also try to figure out what is with stage.

Doest it happen only for some files and not for the others?
 
Could you please select a problematic file and check also .load(stream=True)?
Then could you post please the path from .stage(), the result of .stage().exists() and the path from .path.

**Comment by Koncopd** [comment]
We can also have a call to debug if you have time.

**Comment by falexwolf** [comment]
Already had the call, Sergei! We're on it!

**Comment by falexwolf** [comment]
The main bug that was likely causing frustration here should be fixed with: https://github.com/laminlabs/lamindb/pull/1362

**Comment by Koncopd** [comment]
Anyways, just some tips.
If you suspect a corrupted file in the cache, you can always do `.stage().unlink()` and then `.stage()` again to redownload.
the whole cache can be also cleaned from lamin CLI:
`$ lamin cache clear`

**Comment by Zethson** [comment]
I have the impression that this issue is solved. Please feel free to reopen it if it persists.
